typed-racket-lib/typed-racket/typecheck/tc-envops.rkt

```diff
@@ -56,6 +56,18 @@
         [else (let ([flds* (append lhs (cons (make-fld ty* acc-id #f) (cdr rhs)))])
                 (make-Struct nm par flds* proc poly pred))]))]
 
+    ;; class field ops
+    ;;
+    ;; This is a bit of an abuse of filters/objects. Normally, the object in
+    ;; an expression like (car x) is the `x' identifier. However, for an
+    ;; application of a field accessor (accessor obj), the id `accessor` is
+    ;; the object. We need this so that we can refine the *range* of the
+    ;; accessor, not the object's type.
```

Comment:
Just to make sure I understand what's going on here, what appears to be a refinement of the type of a private field is really a refinement of the return type of the accessor function for that field. We can't turn it into a refinement on the type of the argument to the accessor, since that will have some object type that by definition doesn't mention the private field type. Therefore, `FieldPE` refinements are just a marker saying "the thing being refined is a field represented by this accessor function. Furthermore, this is ok because a private field can't be the result of some other accessor (such as `car`).

Is that all correct?